### ISSUE TITLE
Weird behaviour when narrowing constrained TypeVars

### ISSUE THREAD

**Issue by manulari** [microsoft/pyright#9300]
**Bug description**

Pyright does not seem to narrow constrained type variables correctly — except if you are really insistent and ask repeatedly.

**Screenshot**

![Image](https://github.com/user-attachments/assets/02a58499-07df-44db-b458-70ae9b32f9ea)

(I wasn't sure if there is a difference between the `TypeVar` `S` and `T`, or which one I should be using, but they behave the same here.)

**Code for easier copying**

```
from typing import TypeVar, Literal
from typing_extensions import reveal_type

T = TypeVar('T', bound=Literal['a', 'b'])

def f(x : T) -> T:
    reveal_type(x)
    if x == 'a':
        reveal_type(x)
        if x == 'a':
            reveal_type(x)
    return x

S = TypeVar('S', Literal['a'], Literal['b'])

def g(x : S) -> S:
    reveal_type(x)
    if x == 'a':
        reveal_type(x)
        if x == 'a':
            reveal_type(x)
    return x
```

**VS Code extension or command-line**

Running as a VSCodium extension (ms-pyright), extension version v1.1.386.

**Comment by erictraut** [comment]
Pyright's type narrowing logic for the "x == <literal>" type guard pattern does not currently handle these cases. It's unusual to use literals as an upper bound or as value constraints for a TypeVar, which probably explains why this feature hasn't been requested previously. I think it's a reasonable enhancement request, so I've extended the narrowing logic to handle these cases. This will be included in the next release.

**Comment by manulari** [comment]
Thank you, that was quick :)

Does that mean the following will also start working?:

![Image](https://github.com/user-attachments/assets/e1b0506e-62c4-4fbf-8226-839d0120a341)

(Without the double conditional of course.)

This was where this originally popped up for me. I was trying to use string literals to tag variants of a type that I wanted to distinguish at the type level.

Code for easier copying:

```
from typing import TypeVar, Literal, Generic
from typing_extensions import reveal_type
from dataclasses import dataclass

T = TypeVar('T', bound=Literal['a', 'b'])

@dataclass
class C(Generic[T]):
    k : T
    v : int

ca = C('a', 1)
cb = C('b', 2)

def h(x : T) -> C[T]:
    if x == 'a' and x == 'a':
        reveal_type(x)
        return ca
    else:
        reveal_type(x)
        return cb
```

**Comment by erictraut** [comment]
No, that won't type check for reasons that are unrelated to this issue. The type of `ca` is `C[Literal['a']]`, which isn't assignable to the return type `C[T]`. Type narrowing affects only the expressions involved in the type guard, not other expressions. While you have narrowed the type of `x` to `Literal['a']]` in the first branch, that doesn't change the type of `ca`. It remains unassignable to `C[T]`.

An overload is the appropriate tool to use for this code pattern.

```python
@overload
def h(x: Literal["a"]) -> C[Literal["a"]]: ...
@overload
def h(x: Literal["b"]) -> C[Literal["b"]]: ...
def h(x: Literal["a", "b"]) -> C[Literal["a"]] | C[Literal["b"]]:
    if x == "a":
        reveal_type(x)
        return ca
    else:
        reveal_type(x)
        return cb
```

**Comment by manulari** [comment]
If I replace Literal['a'] with a class A (and similarly with 'b'), this does work:

![Image](https://github.com/user-attachments/assets/40b05b4c-5d80-4657-815b-f2f079ef265f)

Trying to figure out what makes it work, I assumed this had something to do with conditional types? Is there any documentation on how this is supposed to work, beyond the short paragraph [here](https://github.com/microsoft/pyright/blob/main/docs/type-concepts-advanced.md#conditional-types-and-type-variables)?

Code for copying:

```python
class A: pass
class B: pass

R = TypeVar('R', A, B)

@dataclass
class D(Generic[R]):
    k : R
    v : int

da = D(A(), 3)
db = D(B(), 2)

def k(x : R) -> D[R]:
    reveal_type(x)
    if isinstance(x, A):
        reveal_type(x)
        return da
    else:
        reveal_type(x)
        return db
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.387.
